### DISCUSSION TITLE
max_num_elements

### DISCUSSION THREAD

**Discussion by erikhuck** [pyg-team/pytorch_geometric#8010] [Q&A] [answered]
Several aggregation operators have this `max_num_elements` parameter with a definition of "The maximum number of elements within a single aggregation group". What is an "aggregation group"? A graph? What is an "element" in this context? A node?

**Comment by rusty1s** [comment]
Since `Aggregation` can be performed on all kinds of level, e.g., going from edge-level representations to node-level ones, or going from node-level representation to graph-level ones, we framed this rather generally.

Here, `max_num_elements` means the maximum size of elements aggregated into one group. For example, if you have a feature matrix with 6 elements, and `index` looks like `[0, 0, 1, 1, 1, 2]`, then `max_num_elements` would be `index.bincount().max() == 3`.

**Comment by erikhuck** [comment]
@rusty1s In your example, there is a graph with 2 nodes, a graph with 3 nodes, and a graph with 1 node. All these graphs have a different number of nodes. My concern is that the AttentionalAggregation will exclude some of the nodes if the number of nodes surpasses this `max_num_elements` value. But it sounds like, in the case of the AttentionalAggregation, the `max_num_elements` does not have to do with the number of nodes in a graph, but the number of graphs?? What happens when `max_num_elements` is not specified?

**Reply by rusty1s** [reply, marked as answer]
If you do not specify `max_num_elements` it will be inferred via `bincount().max()`, and it will not drop any nodes. It will only drop nodes if you explicitly specify `max_num_elements` to be smaller than the maximum nodes in one of your graphs.

In general, you don't need to worry about setting this argument at all. The reason we introduced this argument in the first-place is due to compilation and to avoid a device sync between CPU and GPU.

**Reply by erikhuck** [reply]
@rusty1s Perfect! Related to this, I see this warning in other aggregation operators, e.g. the LSTMAggregation:
"[LSTMAggregation](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.aggr.LSTMAggregation.html#torch_geometric.nn.aggr.LSTMAggregation) is not a permutation-invariant operator." Do I understand correctly that this is said because the LSTM expects a certain number of elements (i.e. graph nodes) and some graphs have different numbers of nodes? In other words, all the graphs need to have the same number of nodes for the LSTMAggregation or nodes will be excluded? Can the AttentionalAggregation handle any number of graph nodes?

**Reply by rusty1s** [reply]
All aggregations can handle a varying number of nodes. The note refers to that within a group/neighborhood, the aggregation is not permutation-invariant, so you will get different outputs depending on how your neighborhood is sorted. This is due to the sequential nature of LSTMs. The `AttentionalAggregation` doesn't have this issue.
